### ISSUE TITLE
Support HARA and Other Data Entry

### ISSUE THREAD

**Issue by danyeaw** [gaphor/gaphor#927]
### Is your feature request related to a problem?

As a system engineer, I would like to be able to create modeling objects and create relationships through a table or visual templates for quick data entry.

For example, for the Hazard Analysis and Risk Assessment in the ISO 26262 section of RAAML, it is normally created in a table. Each row of the table represents a Hazardous Event, and each column is a data related to the hazardous event, like the Exposure or Situation.

| Name                              | Exposure | Justification Of Exposure                            | Vehicle Usage    | Traffic And People | Location           | Road Condition     | Environmental Condition     |
|-----------------------------------|----------|------------------------------------------------------|------------------|--------------------|--------------------|--------------------|-----------------------------|
| Highway Driving Straight at Speed | E4       | Everyday exposure to city, roads, highways, freeways | Driving at Speed | Traffic Free Flow  | City Roads Highway | Any Road Condition | Any Environmental Condition |
|                                   |          |                                                      |                  |                    |                    |                    |                             |
|                                   |          |                                                      |                  |                    |                    |                    |                             |

If we want to allow for tables, then the creation of a row would create all the objects and relations, which can then be renamed by editing the text in the cells.

Another option is a visual HARA, where all the objects are configured in the same place on the screen like a template. Then, when you create a new Hazardous Event, it creates a whole new diagram with elements and relations. You can then edit each object in the diagram. To make this efficient, we would need to be able to group the HARA diagrams and allow the user the option to go to the next or previous one, and a button to add another one. Here is an example of a single Hazardous Event shown in a BDD.

![image](https://user-images.githubusercontent.com/10014976/130000377-86c24efe-88b0-4d07-b437-192af835bcf2.png)

### Describe the solution you'd like
If we go with the table option, I think there are two main options:

1. Internal table support in Gaphor using a Gtk.TreeView widget.
2. Connector to an external tool like Excel, LibreOffice, or Gnumeric.

An advantage of the internal support is it doesn't require the user to install software and would be faster for quick edits. The external tool would enable all the power of a full spreadsheet, including duplicating cells, adding comments, and easy copying or importing data. We would also only be creating the plumbing to another tool.

Another thing to consider is if we want to allow the user to create generic tables to define what each row and column represents, or hard code each one based on use cases? Also, for a visual template, do we want to allow users to create their own template or hard code each one? 

One of the goals in Gaphor was to keep things simple and everything is done on diagrams themselves, how will this data entry use case fit in with that?

### Describe alternatives you've considered
We could keep the status quo and require users to fill out a normal diagram.

**Comment by danyeaw** [comment]
Here are some examples of how the other tools do it.

### Cameo Safety and Reliability Analyzer Plugin for MagicDraw
Table View
![image](https://user-images.githubusercontent.com/10014976/143246083-99d24d94-c548-4416-8ee6-6aac3b04a4ff.png)

Nodes and Edges (called a Relation Map in MagicDraw)
![image](https://user-images.githubusercontent.com/10014976/143246298-08d2826f-6a2d-40b1-a710-10038738e041.png)

### Ansys medini analyze
![image](https://user-images.githubusercontent.com/10014976/143246641-1b37fb26-b187-4770-8f9f-7749b82fd57b.png)

### Astah Systems Safety
![image](https://user-images.githubusercontent.com/10014976/143246952-a6abd971-87cd-4e4c-898e-92919cac79e6.png)

I will also try to markup a visual HARA example, I don't know of any tools with this capability.

**Comment by amolenaar** [comment]
@danyeaw Do we want to support a tabular format in Gaphor? Are there more applications for tables? I can imagine it would be useful for SysML requirements as well (?)

**Comment by danyeaw** [comment]
I think a tabular diagram would be good for the examples above and for other things like a table of requirements or relations to manage in one place. I am a little torn by this because I think so far we have focused only on putting all elements on a normal diagram.

I did mess around with creating a TreeView in GTK:
![image](https://user-images.githubusercontent.com/10014976/143374816-2317572f-5303-4ccb-819c-fd58cdebae17.png)

**Comment by danyeaw** [comment]
I have been looking at https://github.com/fpaquet/gtksheet which used to be part of a larger and now unmaintained library called GtkExtra.

Positives:
- It provides a spreadsheet like widget which would offer a lot nicer functionality than TreeView out of the box.

Negatives:
- I have been successful at compiling a demo written in c, but I haven't been able to get the gobject introspection working yet.
- The maintainer is responsive, but the project isn't very active and there is no GTK4 support yet.
- It isn't packaged with any of the Linux distros or homebrew,

**Comment by amolenaar** [comment]
This GtkSheet code looks quite old: the way the C code is structured, and it does not have any annotations to generate GIR files.

I think it's better to stick with a `Gtk.TreeView` for now, with a `Gtk.ListStore` backend. It will be easier to support and easier to port once we're on GTK4 (although I do not expect this to be any time soon). It also has the editing features we need, and it's consistent with the rest of the UI.

**Comment by danyeaw** [comment]
@amolenaar Yup, I agree that building on a `Gtk.TreeView` will be the way to go. I wanted to explore what was out there though before starting to build something.

**Comment by vanillajonathan** [comment]
For a regular table you should consider using `Gtk.ColumnView` instead of `Gtk.TreeView`. The `Gtk.ColumnView` is new in GTK 4 and more modern and suited to regular  tables. The old `Gtk.TreeView` is for when you need to collapse/expand.
